### PR TITLE
feat: promisify session cache methods

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Promisify `session.getCacheSize()` and `session.clearCache()`.

cc @deepak1556 @zcbenz

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Converted `session.getCacheSize()` and `session.clearCache()` to return promises instead of taking callbacks.

### CODE REVIEW

File: atom/browser/api/atom_api_session.cc [hunk 1]
```diff
@@ -214,54 +214,82 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
...
     const scoped_refptr<net::URLRequestContextGetter>& context_getter,
     Session::CacheAction action,
-    const net::CompletionCallback& callback) {
+    atom::util::Promise promise) {
   auto* request_context = context_getter->GetURLRequestContext();
+
   auto* http_cache = request_context->http_transaction_factory()->GetCache();
   if (!http_cache)
-    RunCallbackInUI<int>(callback, net::ERR_FAILED);
+    RejectPromiseInUI(std::move(promise), net::ERR_FAILED);
```
Comment (deepak1556) [inline atom/browser/api/atom_api_session.cc]: Should break the code path here if there is no `http_cache`.

```
if (!http_cache) {
  ...
  return;
}
```

File: atom/browser/api/atom_api_session.cc [hunk 2]
```diff
@@ -214,54 +214,82 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
...
+
 // Callback of HttpCache::GetBackend.
 void OnGetBackend(disk_cache::Backend** backend_ptr,
                   Session::CacheAction action,
-                  const net::CompletionCallback& callback,
+                  const atom::util::CopyablePromise& promise,
                   int result) {
   if (result != net::OK) {
-    RunCallbackInUI(callback, result);
+    RejectPromiseInUI(promise.GetPromise(), result);
```
Comment (deepak1556) [inline atom/browser/api/atom_api_session.cc]: Current api documentation for `getCacheSize` and `clearCache` doesn't return error codes when the operation failed, we can simply follow that and reject with emtpy values. Or if we want to return error codes then better to have more specific error messages, like

```
std::string err_msg = "Failed to retrieve cache backend : " + net::ErrorToString(result);
RejectPromiseInUI(promise.GetPromise(), err_msg);
```

File: atom/browser/api/atom_api_session.cc [hunk 3]
```diff
@@ -214,54 +214,82 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
...
   } else if (backend_ptr && *backend_ptr) {
     if (action == Session::CacheAction::CLEAR) {
       (*backend_ptr)
```
Comment (deepak1556) [inline atom/browser/api/atom_api_session.cc]: Should check the result of [DoomAllEntries](https://chromium.googlesource.com/chromium/src/+/master/net/disk_cache/disk_cache.h#217) for `net::ERR_IO_PENDING` . 

```
auto success = (*backend_ptr)->DoomAllEntries(
              base::BindOnce(&ResolveOrRejectPromiseInUI, promise));

if (success != net::ERR_IO_PENDING) {
  ResolveOrRejectPromiseInUI(success, std::move(promise.GetPromise()))
}
```


Also it is better to have a helper function to resolve or reject according to net error codes.

```
void ResolveOrRejectPromiseInUI(int net_error, ...) {
  if (net_error != net::OK) {
    RejectPromiseInUI(net::ErrorToString(net_error), std::move(promise))
  } else {
    ResolvePromiseInUI(std::move(promise));
  }  
} 
```

File: atom/browser/api/atom_api_session.cc [hunk 4]
```diff
@@ -214,54 +214,82 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
...
           base::StringToInt(stat.second, &current_size);
-          RunCallbackInUI(callback, current_size);
+          ResolvePromiseInUI(promise.GetPromise(), current_size);
           break;
         }
       }
     }
   } else {
-    RunCallbackInUI<int>(callback, net::ERR_FAILED);
+    RejectPromiseInUI(promise.GetPromise(), net::ERR_FAILED);
```
Comment (deepak1556) [inline atom/browser/api/atom_api_session.cc]: This condition is not required, as the top level `result != net::OK` will cover it.

File: docs/api/session.md [hunk 5]
```diff
@@ -100,12 +100,22 @@ The following methods are available on instances of `Session`:
...
+
+Returns `Promise<Integer>` - the session's current cache size, in bytes.
+
 #### `ses.clearCache(callback)`
 
 * `callback` Function - Called when operation is done.
 
 Clears the session’s HTTP cache.
 
+#### `ses.clearCache(callback)`
```
Comment (miniak) [inline docs/api/session.md]: remove the `callback` argument

**Comment by deepak1556** [comment]
Would like to see this cleaned up after the suggested code style changes in https://github.com/electron/electron/pull/17222

File: docs/api/session.md [hunk 6]
```diff
@@ -97,12 +97,28 @@ The following methods are available on instances of `Session`:
 
 * `callback` Function
   * `size` Integer - Cache size used in bytes.
+  * `error` Integer - The error code correspnding to the failure.
```
Comment (miniak) [inline docs/api/session.md]: typo correspnding -> corresponding

File: docs/api/session.md [hunk 7]
```diff
@@ -97,12 +97,28 @@ The following methods are available on instances of `Session`:
...
+**[Deprecated Soon](promisification.md)**
+
+#### `ses.getCacheSize()`
+
+Returns `Promise<Integer>` - the session's current cache size, in bytes.
+
 #### `ses.clearCache(callback)`
 
 * `callback` Function - Called when operation is done.
+  * `error` Integer - The error code correspnding to the failure.
```
Comment (miniak) [inline docs/api/session.md]: typo correspnding -> corresponding

File: docs/api/session.md [hunk 8]
```diff
@@ -97,12 +97,28 @@ The following methods are available on instances of `Session`:
...
+
+Clears the session’s HTTP cache.
+
+**[Deprecated Soon](promisification.md)**
+
+#### `ses.clearCache()`
+
+Returns `Promise<void>` - resolves when the cache clear operation is complete.
```
Comment (miniak) [inline docs/api/session.md]: how do we get the error code for failures?

Comment (codebytere) [reply docs/api/session.md]: it rejects with the error code @miniak

Comment (miniak) [reply docs/api/session.md]: got it

File: atom/browser/api/atom_api_session.cc [hunk 9]
```diff
@@ -214,54 +214,76 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
...
 void OnGetBackend(disk_cache::Backend** backend_ptr,
                   Session::CacheAction action,
-                  const net::CompletionCallback& callback,
+                  const atom::util::CopyablePromise& promise,
                   int result) {
   if (result != net::OK) {
-    RunCallbackInUI(callback, result);
+    std::string err_msg =
+        "Failed to retrieve cache backend: " + net::ErrorToString(result);
+    util::Promise::RejectPromise(promise.GetPromise(), err_msg);
```
Comment (deepak1556) [inline atom/browser/api/atom_api_session.cc]: `std::move(err_msg)`

File: atom/browser/api/atom_api_session.cc [hunk 10]
```diff
@@ -214,54 +214,76 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
...
+    atom::util::Promise promise) {
   auto* request_context = context_getter->GetURLRequestContext();
+
   auto* http_cache = request_context->http_transaction_factory()->GetCache();
-  if (!http_cache)
-    RunCallbackInUI<int>(callback, net::ERR_FAILED);
+  if (!http_cache) {
+    std::string err_msg =
+        "Failed to retrieve cache: " + net::ErrorToString(net::ERR_FAILED);
+    util::Promise::RejectPromise(std::move(promise), err_msg);
```
Comment (deepak1556) [inline atom/browser/api/atom_api_session.cc]: `std::move(err_msg)`

File: atom/browser/api/atom_api_session.cc [hunk 11]
```diff
@@ -214,54 +214,76 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
...
 template <typename... T>
 void RunCallbackInUI(const base::Callback<void(T...)>& callback, T... result) {
   base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI},
                            base::BindOnce(callback, result...));
 }
 
+void ResolveOrRejectPromiseInUI(atom::util::Promise promise, int net_error) {
+  if (net_error != net::OK) {
+    std::string err_msg =
+        "Failed to retrieve cache backend: " + net::ErrorToString(net_error);
```
Comment (deepak1556) [inline atom/browser/api/atom_api_session.cc]: Having a specific error message inside a generic helper doesn't look good, either have the error message simply

```
std::string err_msg = net::ErrorToString(net_error);
```

or inline this helper as a lamda.

**Pull request merged**